apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -517,22 +527,39 @@ def _canonicalize_schema(
 def _build_tiledb_schema(
     schema: pa.Schema,
     index_column_names: Sequence[str],
+    domains: Optional[Sequence[Optional[Tuple[Any, Any]]]],
     tiledb_create_options: TileDBCreateOptions,
     context: SOMATileDBContext,
 ) -> tiledb.ArraySchema:
     """Converts an Arrow schema into a TileDB ArraySchema for creation."""
+
+    if domains is None:
+        domains = tuple(None for _ in index_column_names)
+    else:
+        ndom = len(domains)
+        nidx = len(index_column_names)
+        if ndom != nidx:
+            raise ValueError(
+                f"if domains are specified, they must have the same length as index_column_names; got {ndom} != {nidx}"
+            )
+
     dims = []
-    for index_column_name in index_column_names:
+    for index_column_name, domain in zip(index_column_names, domains):
         pa_type = schema.field(index_column_name).type
         dtype = _arrow_types.tiledb_type_from_arrow_type(
             pa_type, is_indexed_column=True
         )
-        domain: Tuple[Any, Any]
-        if isinstance(dtype, str):
+
+        if domain is not None:
```

Comment:
I'm not sure the code path correctly handles `bool` typed index columns.  I believe they are unsupported as in index column, but a (superficial) quick skim looks like it will not generate an error.